### PR TITLE
[WIP]Automated Robotic Arm

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
Adds a stationary robotic arm, which can pick up items on one side and move them into a machine or on the ground of the opposite side. It's a Factorio inserter, I'm not going to mince words. The arm shuts off processing when the source turf is empty of items to save on processing time, and is re-awoken when the turf sends the content change signal.

Robotic arms will try to use the held item on any machines* in the target tile, thus emulating how a player would use it (attacking a machine with metal will insert the stack, and so on). This is done via an attack chain and uses a dummy simplemob to pass as the user, so if there is a machine that cannot accept the item the arm will hit it. If the object is still held after the attack (or if no machine was found), the object is simply dropped on the tile.

*not including atmos pipes and related; if multiple machines are somehow on one tile, only one will be selected.

The Automated Robotic Arm is constructed by making a base using the metal rods crafting window, then adding a metal tile and a borg arm. It is deconstructed by using a screwdriver to remove the arm, a crowbar to remove the tile, and a screwdriver again to return the base to the stack of rods. If you used a left cyborg arm, you will get a left arm back. The Automated Robotic Arm can be rotated while unanchored using alt+click (same as emitters), and a wrench is used to secure or unsecure it.

Also added a signal sent by turfs when their contents change. I am told signals are free, so hopefully every turf in the game calling a signal upon content change proves that.

## Why It's Good For The Game
Adds a bit of automation to the game. Primary use I foresee is in the chem factory, moving pills from a pill stamper to a chem fridge. However, just about any time an object can be used on a machine, an automated arm can be used instead. 

## Changelog
:cl:
add: Automated Robotic Arms are now available for construction. Create the base from any protolathe once the tech is researched, then add a borg arm and secure the result with a wrench. It can insert items into most machines that will accept the item. Otherwise, they can attempt to use the item on mobs, or other items, and if nothing else will drop the item on the ground.
/:cl:

TODO:
- [x] Add in sounds and chat text for various tool (de)construction steps
- [x] Possibly a wire panel for wires that enable/disable upon pulse, and maybe one that _outputs_ a pulse when an item is moved.
- [ ] Attempt some sort of method to show the sprite of the item being moved.
- [x] Add to techwebs, re-do how they are made.
- [ ] Hopefully re-do these awful sprites.

The Automated Robotic Arm, as seen in its intended habitat. In this setup, the arm happens to match the output rate of the pill press exactly.
ezgif-1-259001c3f336

Imagine what could have been.
ezgif-1-f3d447322d49

### CODE REVIEW

**Comment by mrhugo13** [comment]
You are just doing this so you can RP factorio in SS13, aren't you?

**Comment by zxaber** [comment]
Maybe.

But also we can't very well have a chem factory without automated product storage, now can we?

**Comment by mrhugo13** [comment]
Fair point ~~mister factorio man~~

**Comment by Tharcoonvagh** [comment]
Automatic roboarm?
( ͡° ͜ʖ ͡°)

**Comment by GuyonBroadway** [comment]
Drop c4 next to robot arm and giggle as you can blow up walls without admins being able to see the logs

**Comment by BadSS13Player** [comment]
Make an emagged one table people.

**Comment by TheVekter** [comment]
I love it. Except for the sprite, but you know it's awful, so.

**Comment by Fhaxaris** [comment]
imagine all the crazy shit you could do if you could select stuff other than machines.

**Comment by SpinnerMaster** [comment]
Any idea how we could get it to put a BoH into a BoH?

File: code/game/machinery/robotic_arm.dm [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	if(!anchored)
+		STOP_PROCESSING(SSmachines, src)
+	else
+		set_turfs()
+		START_PROCESSING(SSmachines, src)
+
+/obj/machinery/robotic_arm/screwdriver_act(mob/living/user, obj/item/I)
+	if(I.use_tool(src, user, 5, volume=50))
+		if(r_arm)
+			new /obj/item/bodypart/r_arm/robot(get_turf(src))
```
Comment (ATH1909) [inline code/game/machinery/robotic_arm.dm]: this can be used to convert prosthetic arms into cyborg arms, no?

**Comment by Ziiro** [comment]
>imagine all the crazy shit you could do if you could select stuff other than machines.

Give me a machine that you can program to use the crafting menu (i.e, mass produce bots) and we're in fucking business.

**Comment by Buggy123** [comment]
This looks really good, though it'd be really nice if we could get blacklist/whitelist filtering.

Also, how does this perform in worst-case scenarios like if someone drops a thousand pills on the ground with a arm trying to pick them up?

**Comment by zxaber** [comment]
Performance shouldn't be too bad in that case. The arm uses the process() proc to go through the source turf's contents list, and breaks out on the first found item (via a return). Two seconds later, it'll start from the top again (because things might have changed).

It might chug if there's somehow a thousand non-item objects on one tile, but I'm not sure how one would go about doing that.

**Comment by Buggy123** [comment]
>It might chug if there's somehow a thousand non-item objects on one tile, but I'm not sure how one would go about doing that.

Autoproccess cloners churning out braindead clones? I've seen it done before for miasma farms/SM memery.

**Comment by zxaber** [comment]
I'm not convinced that it'll be a huge issue, but I can certainly add adminlogs that get sent if the arm is sorting through an abnormally large list of objects.

**Comment by HomocidlePancakes** [comment]
>It might chug if there's somehow a thousand non-item objects on one tile, but I'm not sure how one would go about doing that.

Blööd tomatoes m8.

File: code/game/machinery/robotic_arm.dm [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+	source = get_step(src, turn(dir,-180))
+	RegisterSignal(source, COMSIG_TURF_CONTENTS_CHANGE, .proc/source_update)
+	dest = get_step(src, dir)
+	RegisterSignal(dest, COMSIG_TURF_CONTENTS_CHANGE, .proc/destination_update)
+
+/obj/machinery/robotic_arm/process()
+	..()
+	if(!anchored || moving)
+		return
+	for(var/O in source.contents)
```
Comment (JJRcop) [inline code/game/machinery/robotic_arm.dm]: You could use `locate(/obj/item) in source`

File: code/game/machinery/robotic_arm.dm [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+/obj/machinery/robotic_arm/proc/source_update()
+	START_PROCESSING(SSmachines, src)
+	return
+
+/obj/machinery/robotic_arm/proc/destination_update()
+	if(target && !QDELETED(target) && get_turf(target) == dest)
+		return
+	target = null
+	for(var/M in dest)
```
Comment (optimumtact) [inline code/game/machinery/robotic_arm.dm]: cringed

**Comment by optimumtact** [comment]
this needs to be tied to a techweb

**Comment by zxaber** [comment]
Added wires!
ezgif-1-095077003b71

Wire interactions are as follows:

| Wire Name  | Cut | Mend | Pulse |
| ------------- | ------------- | ------------- | ---|
| Sensor | Arm no longer grabs items automatically | Arm grabs automatically again | Nothing |
| Motor1 | Arm no longer moves | Arm can move once more | Force arm to attempt to move an item, even if Sensor wire is cut |
|Motor2 | Arm will not insert or drop items (though it will attempt to use them still) | Arm is able to insert and drop items again | Nothing |

I'll probably make it so pulsing the Motor2 wire drops any held item on the source tile.

**Comment by BadSS13Player** [comment]
> > It might chug if there's somehow a thousand non-item objects on one tile, but I'm not sure how one would go about doing that.
> 
> Autoproccess cloners churning out braindead clones? I've seen it done before for miasma farms/SM memery.

Aren't a turf's mob contents held in a separate linked list from a turf's obj contents?

File: code/__DEFINES/traits.dm [hunk 4]
```diff
@@ -252,3 +252,4 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define SLEEPING_CARP_TRAIT "sleeping_carp"
 #define MADE_UNCLONEABLE "made-uncloneable"
 #define TIMESTOP_TRAIT "timestop"
+#define HELD_BY_ROBOARM "heldbyroboarm"
```
Comment (BadSS13Player) [inline code/__DEFINES/traits.dm]:
```suggestion
#define HELD_BY_ROBOARM "heldbyroboarm"

```

File: code/datums/wires/robotic_arm.dm [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+			A.grab()
+	. = ..()
+
+/datum/wires/robotic_arm/on_cut(wire, mend)
+	var/obj/machinery/robotic_arm/A = holder
+	switch(wire)
+		if(WIRE_SENSOR)
+			if(mend)
+				START_PROCESSING(SSmachines, A)
+	. = ..()
```
Comment (BadSS13Player) [inline code/datums/wires/robotic_arm.dm]:
```suggestion
	. = ..()

```

**Comment by zxaber** [comment]
Added to tech webs, under the Advanced Engineering node.

**Comment by swindly** [comment]
> Oh, and they also try to use items on mobs and other items now. 

At long last, self-surgery returns.

File: code/game/machinery/robotic_arm.dm [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+		UnregisterSignal(dest, COMSIG_TURF_CONTENTS_CHANGE)
+	source = get_step(src, turn(dir,-180))
+	RegisterSignal(source, COMSIG_TURF_CONTENTS_CHANGE, .proc/source_update)
+	dest = get_step(src, dir)
+
+/obj/machinery/robotic_arm/process()
+	..()
+	if(!anchored || moving)
+		return
+	if(!anchored  || wires.is_cut(WIRE_SENSOR) || wires.is_cut(WIRE_MOTOR1) || !locate(/obj/item) in source) //If we're unanchored, our detection wire is cut, or the arm rotate wire is cut
```
Comment (JJRcop) [inline code/game/machinery/robotic_arm.dm]: Shouldn't `!locate(/obj/item) in source` be replaced with the `!grab()` seen below, and the one below being removed?

Comment (zxaber) [reply code/game/machinery/robotic_arm.dm]: The `grab()` proc returns true even if the claw alread had something (due to the motor2 wire cut meaning the claw can't drop), which was resulting in a claw that was holding an item swinging back and forth even if the source tile was empty. Hence why the check was copied here.

**Comment by ATH1909** [comment]
Can the claw hug people? Can the claw remove power cells from cyborgs?

If not, can you make it so that it can?

**Comment by zxaber** [comment]
In order:
- The claw does not move without holding an object, which then would be used on the target. I could possibly add the option, but I feel like a stationary, automated robotic arm suddenly hugging would be strange.

- The arm will only grab items off the ground. Adding the ability to pull from other atoms, like borgs, would probably be extra spaghetti code, and I'm not really sure what benefit it would bring other than letting borgs change their own power cells. (And lord knows any hint of a flat borg buff will re-ignite calls for nerfs).

- I could probably make it switch to harm intent under certain conditions, but I'm not sure what the real point would be; If you give the arm something like a toolbox, it'll deal damage all the same on help intent. (Though it would also make the hug option from earlier be a punch option, so that would be funny).

File: code/game/machinery/robotic_arm.dm [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+	layer = BELOW_OBJ_LAYER
+	dir = 1
+	anchored = FALSE
+	var/r_arm = TRUE //False if we need to return a left arm upon disassembly
+	var/moving = FALSE
+	var/obj/machinery/target
+	var/turf/source
+	var/turf/dest
+	var/obj/item/held
+	var/mob/living/simple_animal/robotic_arm_dummy/dummy //A dummy mob to use when inserting objects, which is done via attack chain.
```
Comment (optimumtact) [inline code/game/machinery/robotic_arm.dm]: this should be removed before this is removed

**Comment by BadSS13Player** [comment]
> * I could probably make it switch to harm intent under certain conditions, but I'm not sure what the real point would be; If you give the arm something like a toolbox, it'll deal damage all the same on help intent. (Though it would also make the hug option from earlier be a punch option, so that would be funny).

Some interactions between tools and machines are intent-sensitive, like welding tool on airlocks or anything on the Autolathe.

File: code/game/machinery/robotic_arm.dm (hunk 7 above)
Comment (optimumtact) [reply code/game/machinery/robotic_arm.dm]: changed my mind

File: code/game/machinery/robotic_arm.dm [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+		return TRUE
+	if(!I.use_tool(src, user, 5, volume=50))
+		return 
+	anchored = !anchored
+	if(!anchored)
+		STOP_PROCESSING(SSmachines, src)
+	else
+		set_turfs()
+		START_PROCESSING(SSmachines, src)
+	user.visible_message("<span class='notice'>[user.name] [anchored?"unbolts":"bolts down"] the [src].</span>", \
```
Comment (Dennok) [inline code/game/machinery/robotic_arm.dm]: This `anchored?"unbolts":"bolts down"` inverted here.

**Comment by zxaber** [comment]
Cut me some slack, gitbot, it was the holidays and I was busy.

**Comment by Acer202** [comment]
Do you think you'll ever re-visit this? I stilll think about this a lot.

**Pull request closed without merging**